org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/MapExtensions.java

```diff
@@ -110,4 +122,484 @@ public boolean apply(Map.Entry<K, V> input) {
+	 */
+	public static <K, V> Map<K, V> union(Map<? extends K, ? extends V> left, Map<? extends K, ? extends V> right) {
+		return new MergingMap<K, V>(left, right);
+	}
+
+	private static class MergingMap<K, V> implements Map<K, V> {
+
+		private final Map<? extends K, ? extends V> left;
+
+		private final Map<? extends K, ? extends V> filteredLeft;
+
+		private final Map<? extends K, ? extends V> right;
+
+		MergingMap(Map<? extends K, ? extends V> left, Map<? extends K, ? extends V> right) {
+			this.left = left;
+			this.right = right;
+			this.filteredLeft = Maps.filterKeys(this.left, new Predicate<K>() {
+				@SuppressWarnings("synthetic-access")
+				@Override
+				public boolean apply(K input) {
+					return !MergingMap.this.right.containsKey(input);
+				}
+
+			});
+		}
+
+		@Override
+		public int size() {
+			return this.right.size() + this.filteredLeft.size();
+		}
+
+		@Override
+		public boolean isEmpty() {
+			return this.right.isEmpty() && this.left.isEmpty();
+		}
+
+		@Override
+		public boolean containsKey(Object key) {
+			return this.right.containsKey(key) || this.left.containsKey(key);
+		}
+
+		@Override
+		public boolean containsValue(Object value) {
+			if (this.right.containsValue(value)) {
+				return true;
+			}
+			return this.filteredLeft.containsValue(value);
+		}
+
+		@Override
+		public V get(Object key) {
+			if (this.right.containsKey(key)) {
+				return this.right.get(key);
+			}
+			return this.left.get(key);
+		}
+
+		@Override
+		public V put(K key, V value) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public V remove(Object key) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public void putAll(Map<? extends K, ? extends V> map) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public void clear() {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public Set<K> keySet() {
+			return Sets.union(this.right.keySet(), this.filteredLeft.keySet());
+		}
+
+		@Override
+		public Collection<V> values() {
+			return new MergingCollection();
+		}
+
+		@SuppressWarnings({ "unchecked", "rawtypes" })
+		@Override
+		public Set<Entry<K, V>> entrySet() {
+			final Set ks1 = this.right.entrySet();
+			final Set ks2 = this.filteredLeft.entrySet();
+			return Sets.union(ks1, ks2);
+		}
+
+		/**
+		 * Implementation of the collection of values in a map
+		 * that is representing the merging of two maps.
+		 */
+		@SuppressWarnings("synthetic-access")
+		private class MergingCollection implements Collection<V> {
+
+			MergingCollection() {
+				//
+			}
+
+			@Override
+			public int size() {
+				return MergingMap.this.size();
+			}
+
+			@Override
+			public boolean isEmpty() {
+				return MergingMap.this.isEmpty();
+			}
+
+			@Override
+			public boolean contains(Object object) {
+				return MergingMap.this.containsValue(object);
+			}
+
+			@Override
+			public Iterator<V> iterator() {
+				return Iterators.concat(
```

Comment:
This iterator will support a remove operation for the values of the `right` map. I doubt that this is intended.